jax/_src/dispatch.py

```diff
@@ -548,23 +584,42 @@ def _check_special(name, xla_shape, buf):
 def _execute_compiled(name: str, compiled: XlaExecutable,
                       input_handler: Optional[Callable],
                       output_buffer_counts: Optional[Sequence[int]],
-                      result_handlers, kept_var_idx, *args):
+                      result_handlers,
+                      effects: List[core.Effect],
+                      kept_var_idx, *args):
   device, = compiled.local_devices()
   args = input_handler(args) if input_handler else args
   input_bufs_flat = flatten(device_put(x, device) for i, x in enumerate(args)
                             if i in kept_var_idx)
+  if effects:
```

Comment:
Think we could abstract this into a function? Maybe something like:

```python
input_bufs_flat, output_buffer_counts, token_handler = add_tokens(input_bufs_flat, effects)
... execute and unflatten here ...
out_bufs = token_handler(out_bufs)
return tuple(...)
```

Reply:
Actually, we could compute the update to `output_buffer_counts` earlier, in the place where we `partial(_execute_compiled, ...)`, so it need not be an input and output here IIUC.

Reply:
Done!